app/lib/service/openid/jwt.dart

```diff
@@ -123,7 +119,189 @@ Map<String, dynamic> _decodePart(String part) {
+    } on FormatException {
+      rethrow;
+    } catch (e, st) {
+      _logger.warning('Unexpected JWT parser exception.', e, st);
+      throw FormatException('Unexpected value in JWT payload.');
+    }
+  }
+
+  /// Verifies the timestamps with to the provided [now]:
+  /// - [iat] <= [now]
+  /// - [nbf] <= [now]
+  /// - [now] <= [exp]
+  ///
+  /// Returns `false` if the current timestamp is outside of the allowed range.
+  /// If the timestamp is missing, we treat it as if it would allow the current
+  /// time.
+  bool verifyTimestamps([DateTime? now]) {
+    now ??= clock.now();
+    if (iat != null && iat!.isAfter(now)) {
+      return false;
+    }
+    if (nbf != null && nbf!.isAfter(now)) {
+      return false;
+    }
+    if (exp != null && exp!.isBefore(now)) {
+      return false;
+    }
+    return true;
+  }
+}
+
+/// Parsed payload with the payload values GitHub sends with the token.
+class GitHubJwtPayload {
+  /// user controllable URL identifying the intended audience
+  final String aud;
+
+  /// repository for which the action is running
+  final String repository;
+
+  /// owner of the repository
+  final String repositoryOwner;
+
+  /// name of the event that triggered the workflow
+  final String eventName;
+
+  /// git reference (e.g. branch name or tag name)
+  final String ref;
+
+  /// the kind of git reference given (e.g. "branch")
+  final String refType;
+
+  /// name of the environment used by the job
+  final String environment;
+
+  /// The URL used as the `iss` property of JWT payloads.
+  static const _githubIssuerUrl = 'https://token.actions.githubusercontent.com';
+
+  static const _requiredClaims = <String>{
+    // generic claims
+    'iat',
+    'nbf',
+    'exp',
+    'iss',
+    // github-specific claims
+    'aud',
+    'repository',
+    'repository_owner',
+    'event_name',
+    'ref',
+    'ref_type',
+    'environment',
+  };
+
+  GitHubJwtPayload._(Map<String, dynamic> map)
+      : aud = _parseAsString(map, 'aud'),
+        repository = _parseAsString(map, 'repository'),
+        repositoryOwner = _parseAsString(map, 'repository_owner'),
+        eventName = _parseAsString(map, 'event_name'),
+        ref = _parseAsString(map, 'ref'),
+        refType = _parseAsString(map, 'ref_type'),
+        environment = _parseAsString(map, 'environment');
+
+  factory GitHubJwtPayload(JwtPayload payload) {
+    final missing = _requiredClaims.difference(payload.keys.toSet()).sorted();
+    if (missing.isNotEmpty) {
+      throw FormatException(
+          'JWT from Github is missing following claims: ${missing.map((k) => '`$k`').join(', ')}.');
+    }
+    if (payload.iss != _githubIssuerUrl) {
+      throw FormatException(
+          'Unexpected value in payload: `iss`: `${payload.iss}`.');
+    }
+    return GitHubJwtPayload._(payload);
+  }
+
+  static GitHubJwtPayload? tryParse(JwtPayload payload) {
+    try {
+      return GitHubJwtPayload(payload);
+    } on FormatException {
+      rethrow;
```

Comment:
I'm confused, why not;
```suggestion
      return null;
```